File: Antlr/Ejercicio Py/EvalVisitor.py
import math
from ExprVisitor import ExprVisitor
from ExprParser import ExprParser

class EvalVisitor(ExprVisitor):
    def __init__(self):
        self.memory = {}
        self.use_degrees = True  # True = grados, False = radianes

    def visitAssign(self, ctx):
        id_name = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id_name] = value
        return value

    def visitPrintExpr(self, ctx):
        value = self.visit(ctx.expr())
        print(value)
        return 0

    def visitInt(self, ctx):
        return int(ctx.INT().getText())

    def visitDouble(self, ctx):
        return float(ctx.DOUBLE().getText())

    def visitId(self, ctx):
        id_name = ctx.ID().getText()
        return self.memory.get(id_name, 0)

    def visitParens(self, ctx):
        return self.visit(ctx.expr())

    def visitMulDiv(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == ExprParser.MUL:
            return left * right
        else:
            return left / right

    def visitAddSub(self, ctx):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == ExprParser.ADD:
            return left + right
        else:
            return left - right

    def visitFactorial(self, ctx):
        n = int(self.visit(ctx.expr()))
        if n < 0:
            raise Exception("Factorial no definido para negativos")
        result = 1
        for i in range(2, n+1):
            result *= i
        return result

    def visitFunction(self, ctx):
        func = ctx.func().getText()
        val = self.visit(ctx.expr())

        if self.use_degrees and func in ["Sin", "Cos", "Tan"]:
            val = math.radians(val)

        if func == "Sin":
            return math.sin(val)
        elif func == "Cos":
            return math.cos(val)
        elif func == "Tan":
            return math.tan(val)
        elif func == "Sqrt":
            return math.sqrt(val)
        elif func == "Ln":
            return math.log(val)
        elif func == "Log":
            return math.log10(val)
        else:
            raise Exception(f"Función desconocida: {func}")

File: Antlr/Ejercicio/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends ExprBaseVisitor<Double> {
    Map<String, Double> memory = new HashMap<>();
    boolean useDegrees = true; // true = grados, false = radianes

    @Override
    public Double visitAssign(ExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    @Override
    public Double visitPrintExpr(ExprParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }

    @Override
    public Double visitInt(ExprParser.IntContext ctx) {
        return Double.valueOf(ctx.INT().getText());
    }

    @Override
    public Double visitDouble(ExprParser.DoubleContext ctx) {
        return Double.valueOf(ctx.DOUBLE().getText());
    }

    @Override
    public Double visitId(ExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        if (memory.containsKey(id)) return memory.get(id);
        return 0.0;
    }

    @Override
    public Double visitParens(ExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitMulDiv(ExprParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        return ctx.op.getType() == ExprParser.MUL ? left * right : left / right;
    }

    @Override
    public Double visitAddSub(ExprParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        return ctx.op.getType() == ExprParser.ADD ? left + right : left - right;
    }

    @Override
    public Double visitFactorial(ExprParser.FactorialContext ctx) {
        int n = visit(ctx.expr()).intValue();
        return (double) factorial(n);
    }

    @Override
    public Double visitFunction(ExprParser.FunctionContext ctx) {
        String func = ctx.func().getText();
        double val = visit(ctx.expr());

        if (useDegrees && (func.equals("Sin") || func.equals("Cos") || func.equals("Tan"))) {
            val = Math.toRadians(val);
        }

        switch (func) {
            case "Sin": return Math.sin(val);
            case "Cos": return Math.cos(val);
            case "Tan": return Math.tan(val);
            case "Sqrt": return Math.sqrt(val);
            case "Ln": return Math.log(val);
            case "Log": return Math.log10(val);
            default: throw new RuntimeException("Función desconocida: " + func);
        }
    }

    private long factorial(int n) {
        if (n < 0) throw new RuntimeException("Factorial no definido para negativos");
        long result = 1;
        for (int i = 2; i <= n; i++) result *= i;
        return result;
    }
}


